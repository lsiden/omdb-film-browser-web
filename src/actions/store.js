import { createStore, applyMiddleware, compose } from "redux"
import thunk from "redux-thunk"

import Actions from "actions/types"

const initialState = {
  view: Actions.VIEW_FILM_LIST,
}

export function reduce(state = initialState, action) {
  const { type, data } = action
  switch (type) {
    case Actions.VIEW_FILM_LIST:
    case Actions.VIEW_FILM_DETAIL:
    case Actions.UPDATE_FILMS:
    case Actions.UPDATE_FILM_DETAILS:
    case Actions.UPDATE_IS_FETCHING:
      return {
        ...state,
        ...data,
      }
    case Actions.APPEND_FILMS:
      return {
        ...state,
        ...data,
        films: [...state.films, ...data.films],
      }
    default:
      return state
  }
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
const store = createStore(reduce, composeEnhancers(applyMiddleware(thunk)))

export default store
