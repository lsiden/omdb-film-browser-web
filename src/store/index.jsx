import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'

import ActionType from './const'
import reduce from './reducers'

// Action factories

export const setQuery = (query='') => ({
  type: ActionType.SET_QUERY,
  data: { query },
})

export const replaceFilms = (query='', films=[], totalResults=0) => ({
  type: ActionType.REPLACE_FILMS,
  data: { films, query, totalResults, pageNum: 1 },
})

export const appendFilms = (pageNum=1, films=[]) => ({
  type: ActionType.APPEND_FILMS,
  data: { films, pageNum },
})

export const updateFilmDetails = (filmDetails) => ({
  type: ActionType.UPDATE_FILM_DETAILS,
  data: { filmDetails },
})

export const setFetching = (isFetching=false) => ({
  type: ActionType.UPDATE_IS_FETCHING,
  data: { isFetching }
})

/// //////////////
// State Accessors

export function getFilms(state) {
  return Object.values(state.films || {})
}

/// /////////////
// createStore()

const composeEnhancers = (
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
  && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ trace: true, traceLimit: 25 })
) || compose

export default createStore(reduce, composeEnhancers(applyMiddleware(thunk)))
