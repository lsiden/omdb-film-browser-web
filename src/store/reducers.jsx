import ActionTypes from 'store/const'

const initialState = {
  isFetching: false,
  films: [],
  totalResults: 0,
  query: '',
}

/*
 * I discovered that OMDB search can return some duplicate films.
 * This will cause React to throw a warning when it discovers a duplicate key.
 * So we have to guard against duplicate items.
 */
function reduceUniqueFilms(res, film) {
  res[film.imdbID] = film
  return res
}

const reduceNewFilms = (state, data) => ({
  ...state,
  films: (data.films || []).reduce(reduceUniqueFilms, {}),
  totalResults: parseInt(data.totalResults || '0', 10),
})

const reduceMoreFilms = (state, data) => ({
  ...state,
  films: (data.films || []).reduce(reduceUniqueFilms, { ...state.films || {} })
})

export default function (state=initialState, action) {
  const { type, data } = action
  switch (type) {
    case ActionTypes.UPDATE_FILM_DETAILS:
    case ActionTypes.UPDATE_IS_FETCHING:
    case ActionTypes.SET_QUERY:
      return {
        ...state,
        ...data,
      }
    case ActionTypes.REPLACE_FILMS:
      return reduceNewFilms(state, data)
    case ActionTypes.APPEND_FILMS:
      return reduceMoreFilms(state, data)
    default:
      return state
  }
}
