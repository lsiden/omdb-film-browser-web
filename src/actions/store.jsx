import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'

import Actions from 'actions/types'
import { updateNewFilms, appendFilms } from 'actions/films'

const initialState = {
  view: Actions.VIEW_FILM_LIST,
  isFetching: false,
  films: [],
  totalResults: 0,
  query: '',
}

export function reduce(state = initialState, action) {
  const { type, data } = action
  switch (type) {
    case Actions.UPDATE_FILM_DETAILS:
    case Actions.UPDATE_IS_FETCHING:
    case Actions.SET_QUERY:
      return {
        ...state,
        ...data,
      }
    case Actions.UPDATE_FILMS:
      return updateNewFilms(state, data)
    case Actions.APPEND_FILMS:
      return appendFilms(state, data)
    default:
      return state
  }
}

const composeEnhancers = (
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ trace: true, traceLimit: 25 })
  ) || compose
const store = createStore(reduce, composeEnhancers(applyMiddleware(thunk)))

export default store
