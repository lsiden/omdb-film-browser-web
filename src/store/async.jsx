// Remote Actions

import toastr from 'toastr'
// import promiseFinally from 'promise.prototype.finally'

import store from 'store'
import { FETCH_TIMEOUT, OMDB_API_URL, OMDB_API_KEY } from 'omdb_constants'
import {
  replaceFilms,
  appendFilms,
  updateFilmDetails,
  setFetching,
} from '.'

const queryFetchUri = (query) => `${OMDB_API_URL}?apikey=${OMDB_API_KEY}&type=movie&s=${query}`
const pageFetchUri = (query, page) => queryFetchUri(query) + `&page=${page}`
const filmFetchUri = (id) => `${OMDB_API_URL}?apikey=${OMDB_API_KEY}&type=movie&i=${id}&plot=full`

function toJson(res) {
  try {
    return res.json()
  } catch (e) {
    return Promise.reject(e)
  }
}

const onCatch = (e) => {
  console.error(e)
  toastr.error(e, 'An error occured', {
    preventDuplicates: true,
  })
}

// Returns Promise that resolves to fetched result if successful
const promiseFetchOrTimeout = (dispatch, uri, millisec=FETCH_TIMEOUT) => {
  let timeout
  const timeoutPromise = new Promise((_, reject) => {
    timeout = setTimeout(() => {
      reject(
        new Error(
          'Unable to fetch data. Try a different title or try again later.',
        ),
      )
    }, millisec)
  })
  dispatch(setFetching(true))
  return Promise.race([fetch(uri), timeoutPromise])
    .then(toJson)
    .catch(onCatch)
    .finally(() => {
      clearTimeout(timeout)
      dispatch(setFetching(false))
    })
}

export const promiseQueryResults = (query) => (dispatch) => promiseFetchOrTimeout(
  dispatch, queryFetchUri(query)
).then((res={ Search: [], totalResults: 0 }) => dispatch(replaceFilms(
  query, res.Search, res.totalResults
)))

export const promiseQueryPageFetch = () => (dispatch) => {
  const { query='', pageNum=1 } = store.getState()
  const nextPageNum = pageNum + 1
  return promiseFetchOrTimeout(dispatch, pageFetchUri(query, nextPageNum))
    .then((res={ Search: [], totalResults: 0 }) => {
      dispatch(appendFilms(nextPageNum, res.Search))
    })
}

export const promiseFilmDetails = (id) => (dispatch) => promiseFetchOrTimeout(
  dispatch,
  filmFetchUri(id),
).then((res) => dispatch(updateFilmDetails(res)))
