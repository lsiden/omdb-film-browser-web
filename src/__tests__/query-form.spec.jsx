import React from 'react'
import TestRenderer from 'react-test-renderer'
import { Provider } from 'react-redux'

import QueryForm from 'components/query-form'
import store from 'actions/store'

const defaultProps = {
  onChange: () => {},
  dispatchQueryFetch: () => {},
  clearResults: () => {},
}

const createWrapper = () => TestRenderer.create(
  <Provider store={store}>
    <QueryForm {...defaultProps} />
  </Provider>,
)

describe('QueryForm', () => {
  it('renders a Form element', () => {
    expect(createWrapper()).toMatchSnapshot()
  })
})
