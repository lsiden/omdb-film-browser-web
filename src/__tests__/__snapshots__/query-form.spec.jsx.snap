// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryForm renders a Form element 1`] = `
<form
  style={
    Object {
      "background": "linear-gradient(to right, #000040, #3b00ff)",
      "color": "white",
      "padding": "10px",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <label
      htmlFor="query-form-search-input"
      style={
        Object {
          "marginRight": "1em",
        }
      }
    >
      Search
    </label>
    <div
      style={
        Object {
          "position": "relative",
          "width": "100%",
        }
      }
    >
      <span
        charSet="utf-8"
        dangerouslySetInnerHTML={
          Object {
            "__html": "⚲",
          }
        }
        style={
          Object {
            "color": "#888",
            "fontSize": "20pt",
            "left": "6pt",
            "position": "absolute",
            "top": "-4pt",
            "transform": "rotate(-45deg)",
          }
        }
      />
      <input
        onChange={[Function]}
        placeholder="Film Title ..."
        style={
          Object {
            "color": "black",
            "fontSize": "12pt",
            "height": "24pt",
            "textIndent": "24pt",
            "width": "100%",
          }
        }
        type="search"
        value=""
      />
      <span
        aria-label="Cancel"
        charSet="utf-8"
        className="hover-pointer"
        dangerouslySetInnerHTML={
          Object {
            "__html": "❌",
          }
        }
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        style={
          Object {
            "color": "#888",
            "fontSize": "8pt",
            "position": "absolute",
            "right": 6,
            "top": 6,
          }
        }
        tabIndex={0}
      />
    </div>
  </div>
</form>
`;
